### PR TITLE
Feature/session manager

### PR DESCRIPTION
**Pull request by ryantwarner**
Session manager skeleton

### CODE REVIEW

**Comment by unity-cla-assistant** [comment]
[CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/com.unity.multiplayer.samples.coop?pullRequest=361) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/com.unity.multiplayer.samples.coop?pullRequest=361) before we can accept your contribution.<br/>**2** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: dwoodruffsf<br/>:white_check_mark: eheimburg<br/>:x: lkrell<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.cds.internal.unity3d.com/check/Unity-Technologies/com.unity.multiplayer.samples.coop?pullRequest=361) it.</sub>

File: Assets/BossRoom/Scripts/Client/Net/ClientSessionManager.cs
Comment (SamuelBellomo) [inline Assets/BossRoom/Scripts/Client/Net/ClientSessionManager.cs:43]: Issue with ClientIDs is that they are reused if players reconnect. So clientA and clientB could have client IDs 1 and 2. if both disconnect and reconnect, clientA could have id 2 and clientB could have id 1.
We already have a session guid used to save your character class, I would have consolidated those into your current solution (which seems cleaner :) )

Comment (ryantwarner) [reply Assets/BossRoom/Scripts/Client/Net/ClientSessionManager.cs:43]: I'm not sure how to get the clientGUIDs, I tried changing it to OnConnectionApproval and processing the payload, but I'm not sure how to get that info.

File: Assets/BossRoom/Scripts/Client/Net/ClientSessionManager.cs
Comment (SamuelBellomo) [inline Assets/BossRoom/Scripts/Client/Net/ClientSessionManager.cs:17]: looking at the empty ClientSceneChanged, does this class do anything anymore?

File: Assets/BossRoom/Scripts/Server/Net/ServerSessionManager.cs
Comment (SamuelBellomo) [inline Assets/BossRoom/Scripts/Server/Net/ServerSessionManager.cs:173]: wait, there's two ApprovalChecks?
mmm looking further, looks like callback isn't used. So basically, we're using this one here to insert ourselves during the connection approval process? Isn't there another callback we could use for this, like OnClientConnectedCallback?

File: Assets/BossRoom/Scripts/Server/Net/ServerSessionManager.cs
Comment (SamuelBellomo) [inline Assets/BossRoom/Scripts/Server/Net/ServerSessionManager.cs:201]: beside for checking for success, looks like we're not using that status for anything else?

File: Assets/BossRoom/Scripts/Server/Net/ServerSessionManager.cs
Comment (SamuelBellomo) [inline Assets/BossRoom/Scripts/Server/Net/ServerSessionManager.cs:24]: tested by trying to connect client with host, then having the client disconnect and reconnect.
class was the right one (instead of the default tank), but position was reset to spawn point.
not sure about the name since we seem to have a bug on our hud preventing the display of the name :(

File: Assets/BossRoom/Scripts/Shared/Game/Entity/NetworkCharacterState.cs
Comment (fernando-cortez) [inline Assets/BossRoom/Scripts/Shared/Game/Entity/NetworkCharacterState.cs:117]: We should have the health initialized at the same place in code. Since we're doing that now for the late-joined player inside ServerBossRoomState, we might as well just do this there for a fresh player.

File: Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/ServerGameNetPortal.cs
Comment (fernando-cortez) [inline Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/ServerGameNetPortal.cs:24]: This can be completely removed now right?

File: Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs
Comment (fernando-cortez) [inline Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs:67]: Seems like this was brought over from ServerGameNetPortal and could be removed.

File: Assets/BossRoom/Scripts/Server/Game/Character/PlayerServerCharacter.cs
Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Server/Game/Character/PlayerServerCharacter.cs:43]: I don't think so. The data is saved by a Vec3, but could be by a Quaternion too. I'll change it

File: Assets/BossRoom/Scripts/Server/Game/State/ServerBossRoomState.cs
Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Server/Game/State/ServerBossRoomState.cs:185]: Actually, on the SessionManager's side, it uses the clientID to get the guid, for which the mapping is updated each time a new client connects. But we could skip that step and simply get the guid here instead via ClientPrefs and pass that to the SessionManager

File: Assets/BossRoom/Scripts/Server/Game/State/ServerBossRoomState.cs
Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Server/Game/State/ServerBossRoomState.cs:193]: You're right I think I can move the random avatar assignment to the SessionManager.

File: Assets/BossRoom/Scripts/Shared/Game/Entity/NetworkCharacterState.cs
Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Shared/Game/Entity/NetworkCharacterState.cs:117]: I thought about it too, but this code is used for NPCs as well. So this is needed for them. I'll try to see if I can do the reverse and get the info from the SessionManager here and remove the setting of HP in ServerBossRoomState

Comment (fernando-cortez) [reply Assets/BossRoom/Scripts/Shared/Game/Entity/NetworkCharacterState.cs:117]: That's very true. Hopefully that works out though.

File: Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs
Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs:67]: Yes you're right, I missed this one, thanks

File: Assets/BossRoom/Scripts/Server/Game/State/ServerBossRoomState.cs
Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Server/Game/State/ServerBossRoomState.cs:182]: Nevermind, it doesn't work that way since this code is executed on the host, so we don't have access to the guid through ClientPrefs, and will have to get it via the client ID anyways. Anyways this is fine since the mapping of client IDs and GUIDs is handled by the SessionManager. I just tested it with 2 clients, one via parrelsync and one via a build (I had to revert commit 7315fae to be able to test this) and no matter the order of connection/disconnection/reconnection they always reconnect as the right character.
However I also tested it with 2 clients via parrelsync and there was an issue, the modified guid (adding _Secondary) is not handled correctly and if both clients disconnect and reconnect, they currently both reconnect as the first client's character. I'll look into this.

File: Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs
Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs:113]: The final updates to the sessionPlayerData come in the PlayerServerCharacter's and PersistentPlayer's OnDespawn event, which are called after this.

File: Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/ServerGameNetPortal.cs
Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/ServerGameNetPortal.cs:158]: I'll move this back to the GameNetPortal then, and use the number of connected clients from NetManager instead.

File: Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs
Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs:379]: I thought it was the session manager's responsibility to handle player data that it wants to keep and to reassign, but maybe it could simply carry the information instead. I suppose that I could remove these and make the PersistentPlayer assign them to itself, but I would need to make sure that it does so after the session manager has correctly updated its data.

File: Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs
Comment (SamuelBellomo) [inline Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs:3]: Linq will allocate a bunch and is usually not friendly to game dev. Any reason we're using it?

Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs:3]: It is for this check `m_Portal.NetManager.ConnectedClientsIds.Contains(id)` in the `OnGameEnded` method. I'll look for an alternative

File: Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs
Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs]: In the end, I made it a bit more generic and reusable the only thing that would need to change to reuse it in another project would be the struct holding the data. In parallel I worked on a refactor abstracting all session data using generic types, but the refactored version still has some issues and would still need some work. I think this current version is sufficient for the moment, and would prefer to have this one reviewed and then merge it after eventual new comments are addressed. Then later if we deem it worth it I'll work on ironing out the refactored version and create a new PR for it.

Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs]: Finally the refactored version was easier to iron out than expected, so I pushed it to this branch instead

**Review by fernando-cortez — COMMENTED** [review]
Nitpicking, but I would zero out the position of the session manager prefab and in scene. Will approve once that's updated.

@SamuelBellomo I think your comments have been addressed too.

**Review by fernando-cortez — DISMISSED** [review]

File: Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs
Comment (SamuelBellomo) [inline Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs:15]: Any reason why this is a Monobehaviour? Having this as a pure C# class would make this way easier to manipulate

File: Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs
Comment (SamuelBellomo) [inline Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs:81]: T --> let's use complete names please

File: Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs
Comment (SamuelBellomo) [inline Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs:67]: magic string "host_guid" should be a const somewhere

File: Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/GameNetPortal.cs
Comment (SamuelBellomo) [inline Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/GameNetPortal.cs:177]: why do we need to add a random selection on start host? Can't we wait for the player to select once in the char select scene?

File: Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs
Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs:15]: It is currently placed as a script in a Gameobject, and uses the Awake(), Start() and OnDestroy() events. It also uses the "enabled" field to disable itself on clients. We could make it not a Monobehavior, and have BossRoomSessionManager have this as a field instead of inheriting from it, and refactor it to make it work, but I'm not sure if it would really make it better. I think it would be better to keep most of the logic in SessionManager so that is is easier to reuse

File: Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/GameNetPortal.cs
Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/GameNetPortal.cs:177]: I added his here mostly to remove the responsibility of adding a random class on late join from BossRoomState. It also avoids [this issue](https://jira.unity3d.com/browse/MTT-1464)

File: Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs
Comment (SamuelBellomo) [reply Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs:15]: Your Awake, Start and OnDestroy become useless once you port sessionManager to a self contained class.
Awake's logic becomes singleton logic, Start logic becomes constructor logic, OnDestroy logic becomes destructor logic.
"Enabled" is only useful if your method has Unity callbacks. Thing is, looking at our logic here, I'm not sure I see how this works (since ServerStartedHandler is registered in Start, which means there's no other monobehaviour callbacks which could benefit from enabling/disabling

File: Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs
Comment (SamuelBellomo) [inline Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs:175]: never used?

File: Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs
Comment (SamuelBellomo) [reply Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs:15]:
```suggestion
    // GUID binds a player to a session
    // Once that player connects to a host, the host associates the current ClientID to the player's session GUID.
    // If the player disconnects and reconnects to the same host, the session is preserved.
    public class SessionManager<T> : MonoBehaviour where T : struct, ISessionPlayerData
```

Comment (SamuelBellomo) [reply Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs:15]: Does that summarize it?

Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs:18]: Oh you're right, it doesn't even need to be inside a gameobject, I didn't think about that

File: Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs
Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs:178]: indeed, not used anymore, will remove it

File: Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs
Comment (SamuelBellomo) [inline Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs:249]: "OnSessionEnded"? Just so are consistent in our API

File: Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs
Comment (SamuelBellomo) [inline Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs:20]:
```suggestion
```

Comment (SamuelBellomo) [reply Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/SessionManager.cs:20]: I think I hit the "commit" button and you added it as well

**Comment by SamuelBellomo** [comment]
Beside the monobehaviour removal, I'm all good with this!

**Review by fernando-cortez — APPROVED** [review]

**Review by VALERE91Unity — DISMISSED** [review]
That's a really cool feature :) we should extract some of that knowledge and create a bite-size sample about Session Management based on that work @LukeStampfli @Cosmin-B

**Comment by LPLafontaineB** [comment]
Sorry for that last commit dismissing your reviews, there was one last thing that I missed during my last refactor

**Review by VALERE91Unity — APPROVED** [review]

**Review by SamuelBellomo — APPROVED** [review]

**Pull request merged**
